eclass/nodejs.eclass

```
@@ -0,0 +1,305 @@
+#
+#   Create archive in tar:
+#   tar --create --auto-compress --file foo-1-node_modules.tar.xz foo-1/node_modules/
+
+case ${EAPI} in
+8) ;;
+*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+if [[ -z ${_NODEJS_ECLASS} ]]; then
+_NODEJS_ECLASS=1
+
+# @ECLASS_VARIABLE: NODEJS_MANAGEMENT
+# @PRE_INHERIT
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Specify a package management
+# The default is set to "npm".
+: "${NODEJS_MANAGEMENT:=npm}"
+
+# @ECLASS_VARIABLE: NODEJS_FILES
+# @INTERNAL
+# @DESCRIPTION:
+# Files and directories that usually come in a standard NodeJS/npm module.
+NODEJS_FILES="babel.config.js babel.config.json cli.js dist index.js lib node_modules package.json"
+
+# @ECLASS_VARIABLE: NODEJS_EXTRA_FILES
+# @DESCRIPTION:
+# If additional dist files are present in the NodeJS/npm module that are not
+# listed in NODEJS_FILES, then this is the place to put them in.
+# Can be either files, or directories.
+# Example: NODEJS_EXTRA_FILES="rigger.js modules"
+
+case ${NODEJS_MANAGEMENT} in
+npm)
+    BDEPEND+=" net-libs/nodejs[npm]"
+    ;;
+yarn)
+    BDEPEND+=" sys-apps/yarn"
+    ;;
+*)
+    eerror "Unknown value for \${NODEJS_MANAGEMENT}"
+    die "Value ${NODEJS_MANAGEMENT} is not supported"
+    ;;
+esac
+
+# @FUNCTION: nodejs_version
+# @DESCRIPTION:
+# Returns the package version
+nodejs_version() {
+    node -p "require('./package.json').version"
+}
+
+# @FUNCTION: nodejs_package
+# @DESCRIPTION:
+# Returns the package name
+nodejs_package() {
+    node -p "require('./package.json').name"
+}
+
+# @FUNCTION: nodejs_has_test
+# @DESCRIPTION:
+# Returns true if test script exist
+nodejs_has_test() {
+    node -p "if (require('./package.json').scripts.test === undefined) { process.exit(1) }" &>/dev/null
+}
+
+# @FUNCTION: nodejs_has_build
+# @DESCRIPTION:
+# Returns true if build script exist
+nodejs_has_build() {
+    node -p "if (require('./package.json').scripts.build === undefined) { process.exit(1) }" &>/dev/null
```

This will throw if there is no `scripts` key in `package.json`. Maybe that is intentional?

---

I mean if the package does not have a generic nodejs build,
It may be that it has another form of compilation, only that is analyzed here.
I don't know exactly what you mean?

---

A `package.json` could be:

```json
{
  "name": "zz",
  "version": "0.1.0"
}
```

`require('./package.json').scripts.build` will throw instead of returning `undefined` (Node will exit with non-zero). But since you already accept a bad exit code it probably does not matter here.